compliance/collection/nodeinventorizer/caching_inventorizer.go

```diff
@@ -0,0 +1,284 @@
+// A cached version is returned if it exists and is fresh enough.
+// Otherwise, a new scan guarded by a backoff is run.
+func (c *CachingScanner) Scan(nodeName string) (*storage.NodeInventory, error) {
+	// check whether a cached inventory exists that has not exceeded its validity
+	cache := readInventoryWrap(c.inventoryCachePath)
+	if cache != nil && cache.Inventory != nil && cache.ValidUntil.After(time.Now()) {
+		log.Debugf("Using cached node scan (valid until %v)", cache.ValidUntil)
+		return cache.Inventory, nil
+	}
+
+	// check for existing backoff, wait for specified duration if needed, then persist the new backoff duration
+	backoffDuration := c.initialBackoff
+	if cache != nil {
+		backoffDuration = c.validateBackoff(cache.BackoffDuration)
+	}
+
+	if backoffDuration > c.initialBackoff {
+		log.Warnf("Found existing node scan backoff file - last scan may have failed. Waiting %v seconds before retrying", backoffDuration.Seconds())
+		c.backoffWaitCallback(backoffDuration)
+	}
+
+	// Write backoff duration to cache
+	backoff := inventoryWrap{BackoffDuration: c.calcNextBackoff(backoffDuration)}
+	if err := writeInventoryWrap(backoff, c.inventoryCachePath); err != nil {
+		log.Warnf("Error writing node scan backoff file: %v", err)
+	}
+
+	// if no inventory exists, or it is too old, collect a fresh one
+	newInventory, err := collectInventory(nodeName)
+	if err != nil {
+		return nil, err
+	}
+
+	// Write inventory to cache
+	inventory := inventoryWrap{
+		ValidUntil:      time.Now().Add(c.cacheDuration),
+		BackoffDuration: 0,
+		Inventory:       newInventory,
+	}
+	if err := writeInventoryWrap(inventory, c.inventoryCachePath); err != nil {
+		return nil, errors.Wrap(err, "persisting inventory to cache")
+	}
+
+	return newInventory, nil
+}
+
+func (c *CachingScanner) calcNextBackoff(currentBackoff time.Duration) time.Duration {
+	nextBackoffInterval := currentBackoff * backoffMultiplier
+	if nextBackoffInterval > c.maxBackoff {
+		return c.maxBackoff
+	}
+	return nextBackoffInterval
+}
+
+// validateBackoff ensures that a given duration does not exceed the max backoff setting
+func (c *CachingScanner) validateBackoff(backoff time.Duration) time.Duration {
+	if backoff > c.maxBackoff {
+		return c.maxBackoff
+	}
+	return backoff
+}
+
+func readInventoryWrap(path string) *inventoryWrap {
+	cacheContents, err := os.ReadFile(path)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No node scan cache file found, will run a new scan")
+			return nil
+		}
+		log.Warnf("Unable to read node scan cache, will run a new scan. Error: %v", err)
+		return nil
+	}
+
+	// deserialize stored inventory
+	var wrap inventoryWrap
+	if err := json.Unmarshal(cacheContents, &wrap); err != nil {
+		log.Warnf("Unable to load node scan cache contents, will run a new scan. Error: %v", err)
+		return nil
+	}
+	return &wrap
+}
+
+func writeInventoryWrap(w inventoryWrap, path string) error {
+	jsonWrap, err := json.Marshal(&w)
```

Comment:
Here we come to an area where I'm not at all an expert but I feel some danger.
The thing is that `NodeInventory` is proto-generated type, and there exists a dedicated json marshaller for such types. It's `github.com/golang/protobuf/jsonpb`.

What I can't tell is if there's a significant difference between `encoding/json` and `github.com/golang/protobuf/jsonpb`. If there is, it's most likely devil-in-details, and we may face annoying nuances in behavior later on.

Could you please try swap `json.Marshal` with `jsonpb.Marshal` and `json.Unmarshal` with `jsonutil.JSON*ToProto` (from our `pkg/jsonutil/conversion.go`)?

There will be an obvious problem that `inventoryWrap` is not a `proto.Message` because it was not defined in `.proto` file.
I see two ways to deal with that:
1. Serialize `NodeInventory` using `jsonpb` to string and put it as a string in `inventoryWrap`. Then serialize the resulting `inventoryWrap` with `encoding/json`. Not elegant (as any json-in-json) but straightforward and works.
2. Define a proto message for `inventoryWrap`. The hardest part is finding where to put it and describing in comments why it exists and what purpose it serves.